// angular-sortablejs/dist/src/sortablejs.directive.js
"use strict";
var core_1 = require('@angular/core');
var forms_1 = require('@angular/forms');
var sortablejs_module_1 = require('./sortablejs.module');
// Sortable
var Sortable = require('sortablejs');
// original library calls the events in unnatural order
// first the item is added, then removed from the previous array
// this is a temporary event to work this around
// as long as only one sortable takes place at a certain time
// this is enough to have a single `global` event
var onremove;
var SortablejsDirective = (function () {
    function SortablejsDirective(element, zone) {
        this.element = element;
        this.zone = zone;
        this.runInsideAngular = false;
    }
    SortablejsDirective.prototype.ngOnInit = function () {
        var _this = this;
        if (this.runInsideAngular) {
            this._sortable = new Sortable(this.element.nativeElement, this.options);
        }
        else {
            this.zone.runOutsideAngular(function () {
                _this._sortable = new Sortable(_this.element.nativeElement, _this.options);
            });
        }
    };
    SortablejsDirective.prototype.ngOnChanges = function (changes) {
        var _this = this;
        var optionsChange = changes['inputOptions'];
        if (optionsChange && !optionsChange.isFirstChange()) {
            var previousOptions_1 = optionsChange.previousValue;
            var currentOptions_1 = optionsChange.currentValue;
            Object.keys(currentOptions_1).forEach(function (optionName) {
                if (currentOptions_1[optionName] !== previousOptions_1[optionName]) {
                    // use low-level option setter
                    _this._sortable.option(optionName, currentOptions_1[optionName]);
                }
            });
        }
    };
    SortablejsDirective.prototype.ngOnDestroy = function () {
        this._sortable.destroy();
    };
    Object.defineProperty(SortablejsDirective.prototype, "options", {
        get: function () {
            return Object.assign({}, sortablejs_module_1.SortablejsModule._globalOptions, this.inputOptions, this.overridenOptions);
        },
        enumerable: true,
        configurable: true
    });
    SortablejsDirective.prototype.proxyEvent = function (eventName, event) {
        if (this.inputOptions && this.inputOptions[eventName]) {
            this.inputOptions[eventName](event);
        }
    };
    Object.defineProperty(SortablejsDirective.prototype, "bindingEnabled", {
        // returns whether the items are currently set
        get: function () {
            return !!this.items;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SortablejsDirective.prototype, "overridenOptions", {
        get: function () {
            var _this = this;
            // always intercept standard events but act only in case items are set (bindingEnabled)
            // allows to forget about tracking this.items changes
            return {
                onAdd: function (event) {
                    if (_this.bindingEnabled) {
                        onremove = function (item) {
                            if (_this.items instanceof forms_1.FormArray) {
                                _this.items.insert(event.newIndex, item);
                            }
                            else {
                                _this.items.splice(event.newIndex, 0, item);
                            }
                        };
                    }
                    _this.proxyEvent('onAdd', event);
                },
                onRemove: function (event) {
                    if (_this.bindingEnabled) {
                        var item = void 0;
                        if (_this.items instanceof forms_1.FormArray) {
                            item = _this.items.at(event.oldIndex);
                            _this.items.removeAt(event.oldIndex);
                        }
                        else {
                            item = _this.items.splice(event.oldIndex, 1)[0];
                        }
                        onremove(item);
                        onremove = null;
                    }
                    _this.proxyEvent('onRemove', event);
                },
                onUpdate: function (event) {
                    if (_this.bindingEnabled) {
                        if (_this.items instanceof forms_1.FormArray) {
                            var relocated = _this.items.at(event.oldIndex);
                            _this.items.removeAt(event.oldIndex);
                            _this.items.insert(event.newIndex, relocated);
                        }
                        else {
                            _this.items.splice(event.newIndex, 0, _this.items.splice(event.oldIndex, 1)[0]);
                        }
                    }
                    _this.proxyEvent('onUpdate', event);
                }
            };
        },
        enumerable: true,
        configurable: true
    });
    SortablejsDirective.decorators = [
        { type: core_1.Directive, args: [{
                    selector: '[sortablejs]'
                },] },
    ];
    /** @nocollapse */
    SortablejsDirective.ctorParameters = function () { return [
        { type: core_1.ElementRef, },
        { type: core_1.NgZone, },
    ]; };
    SortablejsDirective.propDecorators = {
        'items': [{ type: core_1.Input, args: ['sortablejs',] },],
        'inputOptions': [{ type: core_1.Input, args: ['sortablejsOptions',] },],
        'runInsideAngular': [{ type: core_1.Input },],
    };
    return SortablejsDirective;
}());
exports.SortablejsDirective = SortablejsDirective;
